File: java_solutions_21-30/Euler_23.java
/*
/   Project Euler problem #23 Non-abundant sums
/
/   Find the sum of all positive integers which cannot be written as the sum
/   of two abundant numbers.
/
*/

public class Euler_23
{
    public static int totalFactors(int n)
    {
        int factorSum = 0;

        for (int i = 1; i < (n / 2) + 1; i++) {
            if (n % i == 0)
                factorSum += i;
        }

        return factorSum;
    }

    public static void main(String[] args)
    {
        int abundantLimit = 28124;
        int abundantNumbers = 0;
        int[] abundantArray = new int[abundantLimit];

        long nonAbundantable = 0;

        for (int i = 1; i < abundantLimit; i++) {
            int factorSum = totalFactors(i);
            if (factorSum > i)
                abundantArray[abundantNumbers++] = i;
        }

        int[] abundantSums = new int[abundantNumbers * abundantNumbers];

        for (int i = 0; i < abundantSums.length; i++)
            abundantSums[i] = 0;

        for (int i = 0; i < abundantNumbers; i++) { 
            for (int j = 0; j < abundantNumbers; j++) {
                abundantSums[abundantArray[i] + abundantArray[j]] = 
                             abundantArray[i] + abundantArray[j];
            }
        }

        for (int i = 0; i < abundantLimit + 1; i++) {
            if (abundantSums[i] == 0)
                nonAbundantable += i;
        }

        System.out.printf("Answer: %d\n", nonAbundantable);
    }
}

File: Python/py_solutions_21-30/Euler_23.py
# Find the sum of all the positive integers which cannot be written as the sum of two abundant numbers. 

from __future__ import print_function

import timeit

try:
    from itertools import ifilterfalse as filterfalse
except ImportError:
    from itertools import filterfalse

from itertools import combinations_with_replacement

try:
    range = xrange
except NameError:
    pass

start = timeit.default_timer()


def euler_23():
    abundant = (i for i in range(28124) if i < 
                sum(filterfalse(lambda x, i=i: i % x, range(1, i // 2 + 1))))
    abundant_sums = {i for i in combo_gen(abundant)}
    return sum(abundant_sums.symmetric_difference(range(28124)))

def combo_gen(x):
     for i in combinations_with_replacement(x, 2):
         if sum(i) < 28124:
            yield sum(i)


print("Answer: {}".format(euler_23()))
stop = timeit.default_timer()
print("Time: {0:9.5f}".format(stop - start))
